### DISCUSSION TITLE
How to read  or interpret the source code in Adonis v5?

### DISCUSSION THREAD

**Discussion by dajpes** [adonisjs/core#1507] [Help] [answered]
Hi, 

So, I wanna try to understand better what Adonis 5 does under the hood, for example:

After installing and make the auth module works, I wanna know how the attempt method truly works, if we go to the docs it says the following:

![image](https://user-images.githubusercontent.com/11248816/90983032-d8665f00-e530-11ea-9e90-a32b99c67d20.png)

Now, if I go to the source code, I find this:

This one points to the guard contract interface
![image](https://user-images.githubusercontent.com/11248816/90983050-ec11c580-e530-11ea-8b66-3cab744e3ec6.png)

This one is used in session:

![image](https://user-images.githubusercontent.com/11248816/90983089-24190880-e531-11ea-995e-47545ab7e7e3.png)

And this one is used in Oat:
![image](https://user-images.githubusercontent.com/11248816/90983096-2da27080-e531-11ea-8cad-3073db932a8e.png)
 

In these 3 examples, let's say I wanna know what the attempt method truly does under the session guard but I don't find anything, I just know it returns a promise and that's pretty much it. Where is the lucid or database query? Where is the actual attempt to login or error thrown in that method?.

Now I'm just talking about the attempt method, but if I wanna look up any other method inside Adonis 5, it's pretty much the same, it just shows what the method returns, but now how it does that. 

I'm new in Typescript, so this could be a big reason why I don't get all of the above.

**Comment by McSneaky** [comment, marked as answer]
There are several methods.
- Read compiled JS (easiest, but might be pain to read)
- Check same methods in Github (requires jumping between editor and Github, quite annoying)
- Clone AdonisJS modules down and install them from local path. It does require the most setup, but then you can actually add some logging or debug code to Adonis modules and see how they behave

**Reply by dajpes** [reply]
The second method worked, now I kinda know how the code works
